python/pyarrow/_compute.pyx

```diff
@@ -2251,3 +2338,219 @@ cdef CExpression _bind(Expression filter, Schema schema) except *:
 
     return GetResultValue(filter.unwrap().Bind(
         deref(pyarrow_unwrap_schema(schema).get())))
+
+
+cdef class ScalarUdfContext:
+    """A container to hold user-defined-function related
+    entities. `batch_length` and `MemoryPool` are important
+    entities in defining functions which require these details. 
+
+    Example
+    -------
+
+    ScalarUdfContext is used with the scalar user-defined-functions. 
+    When defining such a function, the first parameter must be a
+    ScalarUdfContext object. This object can be used to hold important
+    information. This can be further enhanced depending on the use 
+    cases of user-defined-functions. 
+
+    >>> def random(context, one, two):
+            return pc.add(one, two, memory_pool=context.memory_pool)
```

Comment:
Docstrings should have a one-line summary and optionally a description. Also, most of this isn't relevant to a user; I don't think we need to say too much here.

```suggestion
    """
    Per-invocation function context/state.
    
    This object will always be the first argument to a user-defined
    function. It should not be used outside of a call to the function. 
```